{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('..')\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import projgrad\n",
    "plt.style.use('../peptidome.mplstyle')\n",
    "\n",
    "from lib import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "columnsstr = 'seq id> <alignment start> <alignment end> <envelope start> <envelope end> <hmm acc> <hmm name> <type> <hmm start> <hmm end> <hmm length> <bit score> <E-value> <clan'\n",
    "columns = columnsstr.split('> <')\n",
    "df_pfam = pd.read_csv(datadir+'9606_pfam.tsv.gz', sep='\\t', skiprows=3, names=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteome = [(h, seq) for h, seq in fasta_iter(proteome_path('Human'))]\n",
    "headers, seqs = list(zip(*proteome))\n",
    "seqids = [h.split('|')[1] for h in headers]\n",
    "df = pd.DataFrame(dict(seqid=seqids, seq=seqs))\n",
    "seqids_set = set(df['seqid'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2040, 21080, 100194)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(df['seqid'])-set(df_pfam['seq id'])), len(set(df['seqid'])), len(df_pfam['seq id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PF00096    5390\n",
       "PF00400     751\n",
       "PF00028     693\n",
       "PF00041     612\n",
       "PF07679     566\n",
       "Name: hmm acc, dtype: int64"
      ]
     },
     "execution_count": 308,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pfam_filtered = df_pfam[[(s in seqids_set) for s in df_pfam['seq id']]]\n",
    "df_pfam_filtered['hmm acc'].value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq id</th>\n",
       "      <th>alignment start</th>\n",
       "      <th>alignment end</th>\n",
       "      <th>envelope start</th>\n",
       "      <th>envelope end</th>\n",
       "      <th>hmm acc</th>\n",
       "      <th>hmm name</th>\n",
       "      <th>type</th>\n",
       "      <th>hmm start</th>\n",
       "      <th>hmm end</th>\n",
       "      <th>hmm length</th>\n",
       "      <th>bit score</th>\n",
       "      <th>E-value</th>\n",
       "      <th>clan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1319</th>\n",
       "      <td>A0A087WUU8</td>\n",
       "      <td>444</td>\n",
       "      <td>466</td>\n",
       "      <td>444</td>\n",
       "      <td>466</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>28.4</td>\n",
       "      <td>0.0036</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1320</th>\n",
       "      <td>A0A087WUU8</td>\n",
       "      <td>388</td>\n",
       "      <td>410</td>\n",
       "      <td>388</td>\n",
       "      <td>410</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>27.4</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1321</th>\n",
       "      <td>A0A087WUU8</td>\n",
       "      <td>304</td>\n",
       "      <td>326</td>\n",
       "      <td>304</td>\n",
       "      <td>326</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>26.7</td>\n",
       "      <td>0.0120</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1322</th>\n",
       "      <td>A0A087WUU8</td>\n",
       "      <td>276</td>\n",
       "      <td>298</td>\n",
       "      <td>276</td>\n",
       "      <td>298</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>25.6</td>\n",
       "      <td>0.0280</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1323</th>\n",
       "      <td>A0A087WUU8</td>\n",
       "      <td>332</td>\n",
       "      <td>354</td>\n",
       "      <td>332</td>\n",
       "      <td>354</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23.8</td>\n",
       "      <td>0.1000</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seq id  alignment start  alignment end  envelope start  \\\n",
       "1319  A0A087WUU8              444            466             444   \n",
       "1320  A0A087WUU8              388            410             388   \n",
       "1321  A0A087WUU8              304            326             304   \n",
       "1322  A0A087WUU8              276            298             276   \n",
       "1323  A0A087WUU8              332            354             332   \n",
       "\n",
       "      envelope end  hmm acc hmm name  \\\n",
       "1319           466  PF00096  zf-C2H2   \n",
       "1320           410  PF00096  zf-C2H2   \n",
       "1321           326  PF00096  zf-C2H2   \n",
       "1322           298  PF00096  zf-C2H2   \n",
       "1323           354  PF00096  zf-C2H2   \n",
       "\n",
       "                                                   type  hmm start  hmm end  \\\n",
       "1319  PfamLive::Result::SequenceOntology=HASH(0x89b8...          1       23   \n",
       "1320  PfamLive::Result::SequenceOntology=HASH(0x89b8...          1       23   \n",
       "1321  PfamLive::Result::SequenceOntology=HASH(0x89b8...          1       23   \n",
       "1322  PfamLive::Result::SequenceOntology=HASH(0x89b8...          1       23   \n",
       "1323  PfamLive::Result::SequenceOntology=HASH(0x89b8...          1       23   \n",
       "\n",
       "      hmm length  bit score  E-value    clan  \n",
       "1319          23       28.4   0.0036  CL0361  \n",
       "1320          23       27.4   0.0074  CL0361  \n",
       "1321          23       26.7   0.0120  CL0361  \n",
       "1322          23       25.6   0.0280  CL0361  \n",
       "1323          23       23.8   0.1000  CL0361  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zfs = df_pfam_filtered[df_pfam_filtered['hmm acc'] == 'PF00096']\n",
    "zfs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfind = df.set_index('seqid')\n",
    "seq_zf = [dfind.loc[row['seq id']]['seq'][row['envelope start']:row['envelope end']] for ind, row in zfs.iterrows()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['KCKECGKAFKRSSNLTEHRIIH',\n",
       " 'KCEECGKAFNTSSHLTTHKRIH',\n",
       " 'ICEHCGRAFNQSSNLTKHKRIH',\n",
       " 'KCKECGKAFNQSSTLTRHKIIH',\n",
       " 'KCEECGKAFNVSSTLTQHKRIH']"
      ]
     },
     "execution_count": 307,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_zf[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfams_ordered = df_pfam_filtered['hmm acc'].value_counts()\n",
    "pfams_ordered = pfams_ordered.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "multiple_domains = set(pfams_ordered[pfams_ordered['hmm acc'] > 1]['index'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_multiple = df_pfam_filtered[df_pfam_filtered['hmm acc'].isin(multiple_domains)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = []\n",
    "for ind, (seqid, seq) in df.iterrows():\n",
    "    pfam = df_multiple[df_multiple['seq id']==seqid]\n",
    "    if pfam.shape[0] == 0:\n",
    "        seqs.append(seq)\n",
    "    else:\n",
    "        start = 0\n",
    "        pfam = pfam.sort_values('envelope start')\n",
    "        for ind, row in pfam.iterrows():\n",
    "            end = row['envelope start']\n",
    "            seqs.append(seq[start:end])\n",
    "            start = row['envelope end']\n",
    "        end = row['envelope end']\n",
    "        seqs.append(seq[end:])\n",
    "with open('data/human_filtered.txt', 'w') as f:\n",
    "    for seq in seqs:\n",
    "        f.write(seq + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = []\n",
    "for ind, (seqid, seq) in df.iterrows():\n",
    "    pfam = df_multiple[df_multiple['seq id']==seqid]\n",
    "    if pfam.shape[0] == 0:\n",
    "        seqs.append(seq)\n",
    "    else:\n",
    "        newseq = ''\n",
    "        start = 0\n",
    "        pfam = pfam.sort_values('envelope start')\n",
    "        for ind, row in pfam.iterrows():\n",
    "            end = row['envelope start']\n",
    "            newseq += seq[start:end]\n",
    "            newseq += 'X'*(row['envelope end']-row['envelope start'])\n",
    "            start = row['envelope end']\n",
    "        end = row['envelope end']\n",
    "        newseq += seq[end:]\n",
    "        seqs.append(newseq)\n",
    "with open('data/human_replaced.txt', 'w') as f:\n",
    "    for seq in seqs:\n",
    "        f.write(seq + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "zincfingers = set(df_pfam_filtered[df_pfam_filtered['hmm acc'] == 'PF00096']['seq id'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_nozf = df[~df['seqid'].isin(zincfingers)]\n",
    "df_nozf.to_csv('data/human_nozf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfam_ordered_nproteins = df_pfam_filtered.groupby(['seq id', 'hmm acc']).mean().reset_index()['hmm acc'].value_counts()\n",
    "pfam_ordered_nproteins = pfam_ordered_nproteins.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15501"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# keep one random protein per family\n",
    "remove = set()\n",
    "for pfam in pfam_ordered_nproteins[pfam_ordered_nproteins['hmm acc'] > 1]['index']:\n",
    "    family = set(df_pfam_filtered[df_pfam_filtered['hmm acc'] == pfam]['seq id'].unique())\n",
    "    family = family - set(random.sample(family, 1))\n",
    "    remove = remove | family\n",
    "len(remove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_domains = df[~df['seqid'].isin(remove)]\n",
    "df_domains.to_csv('data/human_uniquedomains.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq id</th>\n",
       "      <th>alignment start</th>\n",
       "      <th>alignment end</th>\n",
       "      <th>envelope start</th>\n",
       "      <th>envelope end</th>\n",
       "      <th>hmm acc</th>\n",
       "      <th>hmm name</th>\n",
       "      <th>type</th>\n",
       "      <th>hmm start</th>\n",
       "      <th>hmm end</th>\n",
       "      <th>hmm length</th>\n",
       "      <th>bit score</th>\n",
       "      <th>E-value</th>\n",
       "      <th>clan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16584</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>3</td>\n",
       "      <td>44</td>\n",
       "      <td>3</td>\n",
       "      <td>44</td>\n",
       "      <td>PF01352</td>\n",
       "      <td>KRAB</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0xe0bf...</td>\n",
       "      <td>1</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>78.8</td>\n",
       "      <td>4.200000e-19</td>\n",
       "      <td>No_clan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16596</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>170</td>\n",
       "      <td>192</td>\n",
       "      <td>170</td>\n",
       "      <td>192</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>20.8</td>\n",
       "      <td>9.400000e-01</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16597</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>226</td>\n",
       "      <td>248</td>\n",
       "      <td>226</td>\n",
       "      <td>248</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>20.6</td>\n",
       "      <td>1.100000e+00</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16588</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>282</td>\n",
       "      <td>304</td>\n",
       "      <td>282</td>\n",
       "      <td>304</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>25.2</td>\n",
       "      <td>3.700000e-02</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16587</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>310</td>\n",
       "      <td>332</td>\n",
       "      <td>310</td>\n",
       "      <td>332</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>26.4</td>\n",
       "      <td>1.600000e-02</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16591</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>338</td>\n",
       "      <td>360</td>\n",
       "      <td>338</td>\n",
       "      <td>360</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23.5</td>\n",
       "      <td>1.300000e-01</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16592</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>366</td>\n",
       "      <td>388</td>\n",
       "      <td>366</td>\n",
       "      <td>388</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23.5</td>\n",
       "      <td>1.300000e-01</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16586</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>394</td>\n",
       "      <td>416</td>\n",
       "      <td>394</td>\n",
       "      <td>416</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>27.0</td>\n",
       "      <td>9.700000e-03</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16594</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>422</td>\n",
       "      <td>444</td>\n",
       "      <td>422</td>\n",
       "      <td>444</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>22.0</td>\n",
       "      <td>3.900000e-01</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16585</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>450</td>\n",
       "      <td>472</td>\n",
       "      <td>450</td>\n",
       "      <td>472</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>28.0</td>\n",
       "      <td>4.700000e-03</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16595</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>478</td>\n",
       "      <td>500</td>\n",
       "      <td>478</td>\n",
       "      <td>500</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>21.7</td>\n",
       "      <td>4.900000e-01</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16589</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>506</td>\n",
       "      <td>528</td>\n",
       "      <td>506</td>\n",
       "      <td>528</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>24.5</td>\n",
       "      <td>6.200000e-02</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16593</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>534</td>\n",
       "      <td>556</td>\n",
       "      <td>534</td>\n",
       "      <td>556</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>22.5</td>\n",
       "      <td>2.600000e-01</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16598</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>562</td>\n",
       "      <td>584</td>\n",
       "      <td>562</td>\n",
       "      <td>584</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.500000e+01</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16590</th>\n",
       "      <td>A8MTY0</td>\n",
       "      <td>590</td>\n",
       "      <td>612</td>\n",
       "      <td>590</td>\n",
       "      <td>612</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23.9</td>\n",
       "      <td>9.600000e-02</td>\n",
       "      <td>CL0361</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seq id  alignment start  alignment end  envelope start  envelope end  \\\n",
       "16584  A8MTY0                3             44               3            44   \n",
       "16596  A8MTY0              170            192             170           192   \n",
       "16597  A8MTY0              226            248             226           248   \n",
       "16588  A8MTY0              282            304             282           304   \n",
       "16587  A8MTY0              310            332             310           332   \n",
       "16591  A8MTY0              338            360             338           360   \n",
       "16592  A8MTY0              366            388             366           388   \n",
       "16586  A8MTY0              394            416             394           416   \n",
       "16594  A8MTY0              422            444             422           444   \n",
       "16585  A8MTY0              450            472             450           472   \n",
       "16595  A8MTY0              478            500             478           500   \n",
       "16589  A8MTY0              506            528             506           528   \n",
       "16593  A8MTY0              534            556             534           556   \n",
       "16598  A8MTY0              562            584             562           584   \n",
       "16590  A8MTY0              590            612             590           612   \n",
       "\n",
       "       hmm acc hmm name                                               type  \\\n",
       "16584  PF01352     KRAB  PfamLive::Result::SequenceOntology=HASH(0xe0bf...   \n",
       "16596  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16597  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16588  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16587  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16591  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16592  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16586  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16594  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16585  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16595  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16589  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16593  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16598  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "16590  PF00096  zf-C2H2  PfamLive::Result::SequenceOntology=HASH(0x89b8...   \n",
       "\n",
       "       hmm start  hmm end  hmm length  bit score       E-value     clan  \n",
       "16584          1       42          42       78.8  4.200000e-19  No_clan  \n",
       "16596          1       23          23       20.8  9.400000e-01   CL0361  \n",
       "16597          1       23          23       20.6  1.100000e+00   CL0361  \n",
       "16588          1       23          23       25.2  3.700000e-02   CL0361  \n",
       "16587          1       23          23       26.4  1.600000e-02   CL0361  \n",
       "16591          1       23          23       23.5  1.300000e-01   CL0361  \n",
       "16592          1       23          23       23.5  1.300000e-01   CL0361  \n",
       "16586          1       23          23       27.0  9.700000e-03   CL0361  \n",
       "16594          1       23          23       22.0  3.900000e-01   CL0361  \n",
       "16585          1       23          23       28.0  4.700000e-03   CL0361  \n",
       "16595          1       23          23       21.7  4.900000e-01   CL0361  \n",
       "16589          1       23          23       24.5  6.200000e-02   CL0361  \n",
       "16593          1       23          23       22.5  2.600000e-01   CL0361  \n",
       "16598          1       23          23       17.0  1.500000e+01   CL0361  \n",
       "16590          1       23          23       23.9  9.600000e-02   CL0361  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqid = \"A8MTY0\"\n",
    "seq = dfind.loc['A8MTY0']\n",
    "pfam = df_pfam[df_pfam['seq id']==seqid]\n",
    "pfam.sort_values('envelope start')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Olfactory receptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq id</th>\n",
       "      <th>alignment start</th>\n",
       "      <th>alignment end</th>\n",
       "      <th>envelope start</th>\n",
       "      <th>envelope end</th>\n",
       "      <th>hmm acc</th>\n",
       "      <th>hmm name</th>\n",
       "      <th>type</th>\n",
       "      <th>hmm start</th>\n",
       "      <th>hmm end</th>\n",
       "      <th>hmm length</th>\n",
       "      <th>bit score</th>\n",
       "      <th>E-value</th>\n",
       "      <th>clan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4798</th>\n",
       "      <td>A0A096LPK9</td>\n",
       "      <td>33</td>\n",
       "      <td>303</td>\n",
       "      <td>31</td>\n",
       "      <td>305</td>\n",
       "      <td>PF13853</td>\n",
       "      <td>7tm_4</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>\n",
       "      <td>3</td>\n",
       "      <td>278</td>\n",
       "      <td>280</td>\n",
       "      <td>151.2</td>\n",
       "      <td>1.200000e-40</td>\n",
       "      <td>CL0192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9231</th>\n",
       "      <td>A0A0G2JMP0</td>\n",
       "      <td>32</td>\n",
       "      <td>305</td>\n",
       "      <td>31</td>\n",
       "      <td>307</td>\n",
       "      <td>PF13853</td>\n",
       "      <td>7tm_4</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>\n",
       "      <td>2</td>\n",
       "      <td>278</td>\n",
       "      <td>280</td>\n",
       "      <td>130.1</td>\n",
       "      <td>3.200000e-34</td>\n",
       "      <td>CL0192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9443</th>\n",
       "      <td>A0A0G2JNH3</td>\n",
       "      <td>33</td>\n",
       "      <td>303</td>\n",
       "      <td>31</td>\n",
       "      <td>305</td>\n",
       "      <td>PF13853</td>\n",
       "      <td>7tm_4</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>\n",
       "      <td>3</td>\n",
       "      <td>278</td>\n",
       "      <td>280</td>\n",
       "      <td>147.6</td>\n",
       "      <td>1.500000e-39</td>\n",
       "      <td>CL0192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11021</th>\n",
       "      <td>A0A0X1KG70</td>\n",
       "      <td>32</td>\n",
       "      <td>305</td>\n",
       "      <td>31</td>\n",
       "      <td>307</td>\n",
       "      <td>PF13853</td>\n",
       "      <td>7tm_4</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>\n",
       "      <td>2</td>\n",
       "      <td>278</td>\n",
       "      <td>280</td>\n",
       "      <td>128.8</td>\n",
       "      <td>8.300000e-34</td>\n",
       "      <td>CL0192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14495</th>\n",
       "      <td>A0A286YEU6</td>\n",
       "      <td>34</td>\n",
       "      <td>304</td>\n",
       "      <td>30</td>\n",
       "      <td>307</td>\n",
       "      <td>PF13853</td>\n",
       "      <td>7tm_4</td>\n",
       "      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>\n",
       "      <td>4</td>\n",
       "      <td>276</td>\n",
       "      <td>280</td>\n",
       "      <td>124.2</td>\n",
       "      <td>2.100000e-32</td>\n",
       "      <td>CL0192</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           seq id  alignment start  alignment end  envelope start  \\\n",
       "4798   A0A096LPK9               33            303              31   \n",
       "9231   A0A0G2JMP0               32            305              31   \n",
       "9443   A0A0G2JNH3               33            303              31   \n",
       "11021  A0A0X1KG70               32            305              31   \n",
       "14495  A0A286YEU6               34            304              30   \n",
       "\n",
       "       envelope end  hmm acc hmm name  \\\n",
       "4798            305  PF13853    7tm_4   \n",
       "9231            307  PF13853    7tm_4   \n",
       "9443            305  PF13853    7tm_4   \n",
       "11021           307  PF13853    7tm_4   \n",
       "14495           307  PF13853    7tm_4   \n",
       "\n",
       "                                                    type  hmm start  hmm end  \\\n",
       "4798   PfamLive::Result::SequenceOntology=HASH(0x107e...          3      278   \n",
       "9231   PfamLive::Result::SequenceOntology=HASH(0x107e...          2      278   \n",
       "9443   PfamLive::Result::SequenceOntology=HASH(0x107e...          3      278   \n",
       "11021  PfamLive::Result::SequenceOntology=HASH(0x107e...          2      278   \n",
       "14495  PfamLive::Result::SequenceOntology=HASH(0x107e...          4      276   \n",
       "\n",
       "       hmm length  bit score       E-value    clan  \n",
       "4798          280      151.2  1.200000e-40  CL0192  \n",
       "9231          280      130.1  3.200000e-34  CL0192  \n",
       "9443          280      147.6  1.500000e-39  CL0192  \n",
       "11021         280      128.8  8.300000e-34  CL0192  \n",
       "14495         280      124.2  2.100000e-32  CL0192  "
      ]
     },
     "execution_count": 301,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pfam_filtered[df_pfam_filtered['hmm acc'] == 'PF13853'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = [dfind.loc[row['seq id']]['seq'][row['envelope start']:row['envelope end']]\n",
    "        for ind, row in df_pfam_filtered[df_pfam_filtered['hmm acc'] == 'PF13853'].iterrows()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
