---
layout: post
title: Mutual information as a function of distance
---

An analysis for different proteomes

{% include post-image-gallery.html filter="mivsdist/" %}

### Code 
#### Untitled.ipynb


What is the mutual information between an amino acid and the protein it comes from?


```python
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import sys
sys.path.append('..')
from lib import *
```


```python
data = np.load('../aafreqpca/data/data.npz')
aa_human = data['human']
aa_human = aa_human#[np.random.randint(0, 10000, 5000), :]
```


```python
np.sum(aa_human * np.log2(aa_human/aa_human.mean(axis=0)))/aa_human.shape[0]
```




    0.09741991137385587




```python
meanfreqs.shape
```




    (20, 20)




```python
meanfreq = aa_human.mean(axis=0)
meanfreqs = np.einsum('i,j->ij', meanfreq, meanfreq)
assert meanfreqs[0, 0] == meanfreq[0]**2
```


```python
devs = np.zeros(aa_human.shape[0])
freqss = np.zeros_like(meanfreqs)
for row in range(aa_human.shape[0]):
    freq = aa_human[row]
    freqs = np.einsum('i,j->ij', freq, freq)
    devs[row] = np.sum(freqs * np.log2(freqs/meanfreqs))
    freqss += freqs
freqss /= aa_human.shape[0]
#np.sum(devs)/aa_human.shape[0], np.median(devs)
```

$\sum_i \sum_j < f_i^p f_j^p> \log_2 \frac{< f_i^p f_j^p>}{ <f_i^p> <f_j^p>}$


```python
freqss = np.einsum('ki,kj->ij', aa_human, aa_human)/aa_human.shape[0]
np.sum(freqss * np.log2(freqss/meanfreqs))
```




    0.0014361921190761484




```python
plt.hist(devs, bins=np.linspace(0.0, 1.0, 100));
```


![png](notebook_files/Untitled_9_0.png)



```python
sns.heatmap(np.log2(aa_human/aa_human.mean(axis=0)), vmin=-3, vmax=+3, cmap='PRGn')
#plt.colorbar()
```




    <matplotlib.axes._subplots.AxesSubplot at 0x7f400adf04e0>




![png](notebook_files/Untitled_10_1.png)


$I(X, Z) = H(X) - H(X|Z)$

$I(X_1, X_2) = H(X_2) - H(X_2 | X_1)$
#### run.py

```python
import numpy as np
import pandas as pd

import sys
sys.path.append('..')
from lib import *

name = sys.argv[1]
if name == 'Viruses':
    proteome = datadir + 'human-viruses-uniref90_nohiv.fasta'
else:
    proteome = proteome_path(name)
print(name, proteome)

def calc_mi_std(seqs, gap):
    mis = []
    for i in range(30):
        df2 = Counter(random.sample(seqs, 10000), k=2, gap=gap).to_df(norm=False, clean=True)
        mis.append(calc_mi(df2))
    return np.std(mis, ddof=1)/2**.5

mutualinformation = []
mutualinformation_std = []
shuffled_mutualinformation = []
gaps = np.arange(0, 201, 1)
for gap in gaps:
    seqs = [s for s in fasta_iter(human, returnheader=False)]
    df2 = Counter(seqs, k=2, gap=gap).to_df(norm=False, clean=True)
    mi = calc_mi(df2)
    mutualinformation.append(mi)
    mi_std = calc_mi_std(seqs, gap)
    mutualinformation_std.append(mi_std)

    # calculate shuffled mi
    iterable = scrambled(fasta_iter(proteome, returnheader=False))
    df2 = Counter(iterable, k=2, gap=gap).to_df(norm=False, clean=True)
    shuffledmi = calc_mi(df2)
    shuffled_mutualinformation.append(shuffledmi)
    print(gap, mi, mi_std, shuffledmi)

df = pd.DataFrame.from_dict(dict(gaps=gaps, mutualinformation=mutualinformation,
                                 mutualinformation_std=mutualinformation_std,
                                 shuffledmutualinformation=shuffled_mutualinformation))
df.to_csv('data/mutualinformation-%s.csv'%name)

```
#### plot.py

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('../peptidome.mplstyle')

import sys
sys.path.append('..')
from lib import *

for species in ['Human', 'Mouse', 'Yeast']:
    fig, ax = plt.subplots()
    df = pd.read_csv('data/mutualinformation-%s.csv'%species)
    ax.plot(df['gaps']+1, df['mutualinformation'], label='data')
    ax.plot(df['gaps']+1, df['shuffledmutualinformation'], label='shuffled')
    ax.legend()
    ax.set_ylim(0.0, 0.0145)
    ax.set_xlim(1.0, 200.0)
    ax.set_xscale('log')
    ax.set_xlabel('Distance')
    ax.set_ylabel('Mutual information in bits')
    fig.tight_layout()
    if species == 'Human':
        fig.savefig('main.png')
        plt.show()
    else:
        fig.savefig('%s.png'%species)

```
