---
layout: post
title: How does filtering by netMHC affect likelihoods?
---

In particular for Malaria.

{% include post-image-gallery.html filter="netmhc/" %}

### Code 
#### netmhc.ipynb

# Using netMHC predictions


```python
import glob, json
import numpy as np
import pandas as pd
from scipy.stats import entropy
import sklearn.decomposition
import sklearn.manifold
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

import sys
sys.path.append('..')

from lib import *
from IPython.display import display, HTML
#import weblogolib
#from corebio.seq import SeqList, Alphabet
```


```python
dfbinders = pd.concat(pd.read_csv(filename, sep=',') for filename in glob.glob('/home/amayer/data/peptidome/netmhc/Malaria-HLA-*.csv'))
```

    /home/amayer/.conda/envs/py3/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
    of pandas will change to not sort by default.
    
    To accept the future behavior, pass 'sort=False'.
    
    To retain the current behavior and silence the warning, pass 'sort=True'.
    
      """Entry point for launching an IPython kernel.



```python
totalpeptides = sum(len(seq) for h, seq in fasta_iter(human))
totalpeptides
```




    11518487




```python
counter9 = count_kmers_proteome(malaria, 9)
```


```python
len(counter9)
```




    3811104




```python
len(dfbinders['Peptide'].unique()), len(np.array(dfbinders['Peptide']))
```




    (1433995, 6034826)




```python
# 50nM = strong binding
# 500nM = weak binding
dfbinders = dfbinders[dfbinders['nM']<500]
dfbinders.head()
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>HLA-B5701</th>
      <th>Core</th>
      <th>H_Avg_Ranks</th>
      <th>ID</th>
      <th>N_binders</th>
      <th>Peptide</th>
      <th>Pos</th>
      <th>Rank</th>
      <th>Unnamed: 0</th>
      <th>nM</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>NaN</td>
      <td>FVFILSVLI</td>
      <td>3.00</td>
      <td>sp_C6KT68_FENR_</td>
      <td>0.0</td>
      <td>FVFILSVLI</td>
      <td>4.0</td>
      <td>3.00</td>
      <td>4.0</td>
      <td>260.5</td>
    </tr>
    <tr>
      <th>1</th>
      <td>NaN</td>
      <td>ILSVLISGV</td>
      <td>0.06</td>
      <td>sp_C6KT68_FENR_</td>
      <td>1.0</td>
      <td>ILSVLISGV</td>
      <td>7.0</td>
      <td>0.06</td>
      <td>7.0</td>
      <td>4.3</td>
    </tr>
    <tr>
      <th>2</th>
      <td>NaN</td>
      <td>VLISGVCCI</td>
      <td>1.00</td>
      <td>sp_C6KT68_FENR_</td>
      <td>1.0</td>
      <td>VLISGVCCI</td>
      <td>10.0</td>
      <td>1.00</td>
      <td>10.0</td>
      <td>38.2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>NaN</td>
      <td>RMTAHSRFL</td>
      <td>1.20</td>
      <td>sp_C6KT68_FENR_</td>
      <td>1.0</td>
      <td>RMTAHSRFL</td>
      <td>29.0</td>
      <td>1.20</td>
      <td>29.0</td>
      <td>51.3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>NaN</td>
      <td>NLYTVKNPL</td>
      <td>1.20</td>
      <td>sp_C6KT68_FENR_</td>
      <td>1.0</td>
      <td>NLYTVKNPL</td>
      <td>62.0</td>
      <td>1.20</td>
      <td>62.0</td>
      <td>52.9</td>
    </tr>
  </tbody>
</table>
</div>




```python
for filename in glob.glob('/home/amayer/data/peptidome/netmhc/Human-HLA-*.csv'):
    dfbinders = pd.read_csv(filename, sep=',')
    seqs = SeqList(list(dfbinders['Peptide']), alphabet=Alphabet(aminoacids))
    data = weblogolib.LogoData.from_seqs(seqs)
    options = weblogolib.LogoOptions()
    #options.logo_title = 'HLA-A0101'
    f = weblogolib.LogoFormat(data, options)
    svg = weblogolib.svg_formatter(data, f)
    display(HTML(svg.decode('ascii')))
```


```python
with open('../../data/triplet-human.json', 'r') as f:
    tripletparams = json.load(f)
```


```python
k = 9
loglikelihood = lambda seq, k: loglikelihood_triplet(seq, **tripletparams, k=k)
phuman = np.array([loglikelihood(seq[i:i+k], k) for h, seq in fasta_iter(human) for i in range(len(seq)-k+1) ])
```


```python
dfepitope['Epitope', 'Description']
```




    6830      SLKKNSRSL
    44829     YLNKIQNSL
    44840     KPIVQYDNF
    44872     GLIMVLSFL
    45349     DLLEEGNTL
    45350     KLEELHENV
    45351     VLDKVEETV
    45352     GLLNKLENI
    45353     MEKLKELEK
    45354     EPKDEIVEV
    45428     KPIVQYDNF
    45429     KPIVQYDNF
    45440     LRKPKHKKL
    45442     ASKNKEKAL
    45444     KNKEKALII
    45446     MPNDPNRNV
    45447     MPNDPNRNV
    45449     HLGNVKYLV
    45452     GIAGGLALL
    45453     MINAYLDKL
    45488     ASKNKEKAL
    45489     KNKEKALII
    45490     MPLETQLAI
    45491     MPLETQLAI
    46066     GLIMVLSFL
    46069     YLNKIQNSL
    46072     MPNDPNRNV
    46176     YLNKIQNSL
    46177     LRKPKHKKL
    46179     KILSVFFLA
                ...    
    109315    YLNKIQNSL
    109320    MPNDPNRNV
    170798    FEFTYMINF
    170800    HPKEYEYPL
    170801    YEYPLHQEH
    170805    NEFPAIDLF
    170809    TQKCEIFNV
    170811    NEVVVKEEY
    173526    KPIVQYDNF
    173527    APFISAVAA
    223362    TLDEMRHFY
    223364    NEVVVKEEY
    255837    KSHGKGYNW
    265743    KSHGKGYNW
    303887    FLFVEALFQ
    303889    FVEALFQEY
    303893    IMVLSFLFL
    303895    LIMVLSFLF
    303896    SFLFVEALF
    303897    VFNVVNSSI
    304267    FVEALFQEY
    304269    LIMVLSFLF
    304270    LIMVLSFLF
    304277    SFLFVEALF
    304278    SFLFVEALF
    306820    KSHGKGYNW
    306821    KSRGKGYNW
    307451    KSHGKGYNW
    307452    KSRGKGYNW
    307654    YLNKIQNSL
    Name: (Epitope, Description), Length: 140, dtype: object




```python
dfproteomes = load_proteomes()
pathogenproteomes = dfproteomes[dfproteomes['type'].isin(['parasite'])]

dfepitopes = load_iedb_tcellepitopes(human_only=True, positive_only=True, peptide_length=9)
#dfepitopes = pd.read_csv(datadir+'allhuman-iedb-epitopes.csv', header=1, usecols=(2, 9, 14))
iedbname = pathogenproteomes['iedbname'].iloc[0]
dfepitopes = dfepitopes.fillna('')
dfepitope = dfepitopes[dfepitopes['Epitope', 'Parent Species'].str.contains(iedbname)]
#dfepitope = pd.read_csv(datadir+'iedb-Malaria-MHCI-humanhost.csv', header=1, usecols=(2, 9, 14))
#dfepitope['length'] = [len(d) for d in dfepitope['Description']]
```


```python
len(np.unique(epi))
```




    54




```python
epi = list(dfepitope['Epitope', 'Description'])
epi = np.unique(epi)
pepitope = np.array([loglikelihood(seq, k) for seq in epi])
#epi = list(dfepitope['Description'])
#pepitope = np.array([loglikelihood(seq[i:i+k], k) for seq in epi for i in range(len(seq)-k+1)])
pepitope = pepitope[~np.isnan(pepitope)]

path = datadir + pathogenproteomes['path'].iloc[0]
ppathogen = np.array([loglikelihood(seq[i:i+k], k) for h, seq in fasta_iter(path) for i in range(len(seq)-k+1) ])
ppathogen = ppathogen[~np.isnan(ppathogen)]

print(len(phuman), len(ppathogen), len(pepitope))
```

    11349864 4129446 54



```python
preds = []
for filename in glob.glob('/home/amayer/repos/peptidome/data/netmhc/Malaria-HLA-*.csv'):
    dfbinders = pd.read_csv(filename, sep=',')
    ppredepitope = np.array([loglikelihood(seq, k) for seq in dfbinders['Peptide'].unique()])
    ppredepitope = ppredepitope[~np.isnan(ppredepitope)]
    if len(ppredepitope) > 20:
        preds.append(ppredepitope)
```


```python
for mhc in [False, True]:
    fig, ax = plt.subplots()
    if mhc:
        plot_histograms(preds,
                        [None]*len(preds),
                        nbins=20, lw=.5, color='k', alpha=.5,
                        xmin=-14, xmax=-9, ax=ax)
    plot_histograms([phuman, ppathogen, pepitope],
                    ['human', 'pathogen', 'epitope'],
                    nbins=20, lw=2,
                    xmin=-14, xmax=-9, ax=ax)
    ax.set_ylabel('relative frequency')
    ax.set_ylim(0.0, 0.22)
    ax.set_yticks([])
    ax.set_xlabel('$\log_{10}$ probability given human proteome statistics')
    fig.tight_layout()
    #fig.savefig('plots/likelihoodprofile-Malaria-triplet-k%i%s.png' % (k, '-mhc' if mhc else ''), dpi=300)
    if mhc:
        fig.savefig('main.png')
```

    No handles with labels found to put in legend.



![png](notebook_files/netmhc_16_1.png)



![png](notebook_files/netmhc_16_2.png)



```python
dfepi9 = dfepitope[dfepitope['length'] == 9]
```


```python
len(set(dfepi9['Description'])), len(set(dfepi9['Description']) - set(dfbinders['Peptide'].unique()))
```


```python
with open('peptides.txt', 'w') as f:
    for item in list(dfepi9['Description']):
        f.write("%s\n" % item)
```


```python
!netMHC peptides.txt -xls -xlsfile peptides.csv -a HLA-A0101 -p
```


```python
pd.read_csv('peptides.csv', sep='\t', skiprows=1)
```


```python

```
