---
layout: post
title: How much confounding is there due to shared protein domains?
---

An analysis based on Pfam domain assignments. As expected the dominant peaks in the mutual information vanish once we exclude all proteins containing Zinc finger motifs. We can also use the domain information in various ways to filter the protein data to more generally assess how much shared protein domains affect the mutual information. One sensible way to do so is to keep just one randmly chosen protein per protein domain family. This also removes the Zinc finger peak, but otherwise the decay of mutual information is broadly similar. 

{% include post-image-gallery.html filter="pfam/" %}

### Code 
#### pfam.ipynb

```python
import sys
sys.path.append('..')
import random
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import projgrad
plt.style.use('../peptidome.mplstyle')

from lib import *

```


```python
columnsstr = 'seq id> <alignment start> <alignment end> <envelope start> <envelope end> <hmm acc> <hmm name> <type> <hmm start> <hmm end> <hmm length> <bit score> <E-value> <clan'
columns = columnsstr.split('> <')
df_pfam = pd.read_csv(datadir+'9606_pfam.tsv.gz', sep='\t', skiprows=3, names=columns)
```


```python
proteome = [(h, seq) for h, seq in fasta_iter(proteome_path('Human'))]
headers, seqs = list(zip(*proteome))
seqids = [h.split('|')[1] for h in headers]
df = pd.DataFrame(dict(seqid=seqids, seq=seqs))
seqids_set = set(df['seqid'])
```


```python
len(set(df['seqid'])-set(df_pfam['seq id'])), len(set(df['seqid'])), len(df_pfam['seq id'])
```




    (2040, 21080, 100194)




```python
df_pfam_filtered = df_pfam[[(s in seqids_set) for s in df_pfam['seq id']]]
df_pfam_filtered['hmm acc'].value_counts().head()
```




    PF00096    5390
    PF00400     751
    PF00028     693
    PF00041     612
    PF07679     566
    Name: hmm acc, dtype: int64




```python
zfs = df_pfam_filtered[df_pfam_filtered['hmm acc'] == 'PF00096']
zfs.head()
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seq id</th>
      <th>alignment start</th>
      <th>alignment end</th>
      <th>envelope start</th>
      <th>envelope end</th>
      <th>hmm acc</th>
      <th>hmm name</th>
      <th>type</th>
      <th>hmm start</th>
      <th>hmm end</th>
      <th>hmm length</th>
      <th>bit score</th>
      <th>E-value</th>
      <th>clan</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1319</th>
      <td>A0A087WUU8</td>
      <td>444</td>
      <td>466</td>
      <td>444</td>
      <td>466</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>28.4</td>
      <td>0.0036</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>1320</th>
      <td>A0A087WUU8</td>
      <td>388</td>
      <td>410</td>
      <td>388</td>
      <td>410</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>27.4</td>
      <td>0.0074</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>1321</th>
      <td>A0A087WUU8</td>
      <td>304</td>
      <td>326</td>
      <td>304</td>
      <td>326</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>26.7</td>
      <td>0.0120</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>1322</th>
      <td>A0A087WUU8</td>
      <td>276</td>
      <td>298</td>
      <td>276</td>
      <td>298</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>25.6</td>
      <td>0.0280</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>1323</th>
      <td>A0A087WUU8</td>
      <td>332</td>
      <td>354</td>
      <td>332</td>
      <td>354</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>23.8</td>
      <td>0.1000</td>
      <td>CL0361</td>
    </tr>
  </tbody>
</table>
</div>




```python
dfind = df.set_index('seqid')
seq_zf = [dfind.loc[row['seq id']]['seq'][row['envelope start']:row['envelope end']] for ind, row in zfs.iterrows()]
```


```python
seq_zf[:5]
```




    ['KCKECGKAFKRSSNLTEHRIIH',
     'KCEECGKAFNTSSHLTTHKRIH',
     'ICEHCGRAFNQSSNLTKHKRIH',
     'KCKECGKAFNQSSTLTRHKIIH',
     'KCEECGKAFNVSSTLTQHKRIH']




```python
pfams_ordered = df_pfam_filtered['hmm acc'].value_counts()
pfams_ordered = pfams_ordered.reset_index()
```


```python
multiple_domains = set(pfams_ordered[pfams_ordered['hmm acc'] > 1]['index'])
```


```python
df_multiple = df_pfam_filtered[df_pfam_filtered['hmm acc'].isin(multiple_domains)]
```


```python
seqs = []
for ind, (seqid, seq) in df.iterrows():
    pfam = df_multiple[df_multiple['seq id']==seqid]
    if pfam.shape[0] == 0:
        seqs.append(seq)
    else:
        start = 0
        pfam = pfam.sort_values('envelope start')
        for ind, row in pfam.iterrows():
            end = row['envelope start']
            seqs.append(seq[start:end])
            start = row['envelope end']
        end = row['envelope end']
        seqs.append(seq[end:])
with open('data/human_filtered.txt', 'w') as f:
    for seq in seqs:
        f.write(seq + '\n')
```


```python
seqs = []
for ind, (seqid, seq) in df.iterrows():
    pfam = df_multiple[df_multiple['seq id']==seqid]
    if pfam.shape[0] == 0:
        seqs.append(seq)
    else:
        newseq = ''
        start = 0
        pfam = pfam.sort_values('envelope start')
        for ind, row in pfam.iterrows():
            end = row['envelope start']
            newseq += seq[start:end]
            newseq += 'X'*(row['envelope end']-row['envelope start'])
            start = row['envelope end']
        end = row['envelope end']
        newseq += seq[end:]
        seqs.append(newseq)
with open('data/human_replaced.txt', 'w') as f:
    for seq in seqs:
        f.write(seq + '\n')
```


```python
zincfingers = set(df_pfam_filtered[df_pfam_filtered['hmm acc'] == 'PF00096']['seq id'].unique())
```


```python
df_nozf = df[~df['seqid'].isin(zincfingers)]
df_nozf.to_csv('data/human_nozf.csv')
```


```python
pfam_ordered_nproteins = df_pfam_filtered.groupby(['seq id', 'hmm acc']).mean().reset_index()['hmm acc'].value_counts()
pfam_ordered_nproteins = pfam_ordered_nproteins.reset_index()
```


```python
# keep one random protein per family
remove = set()
for pfam in pfam_ordered_nproteins[pfam_ordered_nproteins['hmm acc'] > 1]['index']:
    family = set(df_pfam_filtered[df_pfam_filtered['hmm acc'] == pfam]['seq id'].unique())
    family = family - set(random.sample(family, 1))
    remove = remove | family
len(remove)
```




    15501




```python
df_domains = df[~df['seqid'].isin(remove)]
df_domains.to_csv('data/human_uniquedomains.csv')
```


```python
seqid = "A8MTY0"
seq = dfind.loc['A8MTY0']
pfam = df_pfam[df_pfam['seq id']==seqid]
pfam.sort_values('envelope start')
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seq id</th>
      <th>alignment start</th>
      <th>alignment end</th>
      <th>envelope start</th>
      <th>envelope end</th>
      <th>hmm acc</th>
      <th>hmm name</th>
      <th>type</th>
      <th>hmm start</th>
      <th>hmm end</th>
      <th>hmm length</th>
      <th>bit score</th>
      <th>E-value</th>
      <th>clan</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>16584</th>
      <td>A8MTY0</td>
      <td>3</td>
      <td>44</td>
      <td>3</td>
      <td>44</td>
      <td>PF01352</td>
      <td>KRAB</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0xe0bf...</td>
      <td>1</td>
      <td>42</td>
      <td>42</td>
      <td>78.8</td>
      <td>4.200000e-19</td>
      <td>No_clan</td>
    </tr>
    <tr>
      <th>16596</th>
      <td>A8MTY0</td>
      <td>170</td>
      <td>192</td>
      <td>170</td>
      <td>192</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>20.8</td>
      <td>9.400000e-01</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16597</th>
      <td>A8MTY0</td>
      <td>226</td>
      <td>248</td>
      <td>226</td>
      <td>248</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>20.6</td>
      <td>1.100000e+00</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16588</th>
      <td>A8MTY0</td>
      <td>282</td>
      <td>304</td>
      <td>282</td>
      <td>304</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>25.2</td>
      <td>3.700000e-02</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16587</th>
      <td>A8MTY0</td>
      <td>310</td>
      <td>332</td>
      <td>310</td>
      <td>332</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>26.4</td>
      <td>1.600000e-02</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16591</th>
      <td>A8MTY0</td>
      <td>338</td>
      <td>360</td>
      <td>338</td>
      <td>360</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>23.5</td>
      <td>1.300000e-01</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16592</th>
      <td>A8MTY0</td>
      <td>366</td>
      <td>388</td>
      <td>366</td>
      <td>388</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>23.5</td>
      <td>1.300000e-01</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16586</th>
      <td>A8MTY0</td>
      <td>394</td>
      <td>416</td>
      <td>394</td>
      <td>416</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>27.0</td>
      <td>9.700000e-03</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16594</th>
      <td>A8MTY0</td>
      <td>422</td>
      <td>444</td>
      <td>422</td>
      <td>444</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>22.0</td>
      <td>3.900000e-01</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16585</th>
      <td>A8MTY0</td>
      <td>450</td>
      <td>472</td>
      <td>450</td>
      <td>472</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>28.0</td>
      <td>4.700000e-03</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16595</th>
      <td>A8MTY0</td>
      <td>478</td>
      <td>500</td>
      <td>478</td>
      <td>500</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>21.7</td>
      <td>4.900000e-01</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16589</th>
      <td>A8MTY0</td>
      <td>506</td>
      <td>528</td>
      <td>506</td>
      <td>528</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>24.5</td>
      <td>6.200000e-02</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16593</th>
      <td>A8MTY0</td>
      <td>534</td>
      <td>556</td>
      <td>534</td>
      <td>556</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>22.5</td>
      <td>2.600000e-01</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16598</th>
      <td>A8MTY0</td>
      <td>562</td>
      <td>584</td>
      <td>562</td>
      <td>584</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>17.0</td>
      <td>1.500000e+01</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16590</th>
      <td>A8MTY0</td>
      <td>590</td>
      <td>612</td>
      <td>590</td>
      <td>612</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>23.9</td>
      <td>9.600000e-02</td>
      <td>CL0361</td>
    </tr>
  </tbody>
</table>
</div>



## Olfactory receptors


```python
df_pfam_filtered[df_pfam_filtered['hmm acc'] == 'PF13853'].head()
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seq id</th>
      <th>alignment start</th>
      <th>alignment end</th>
      <th>envelope start</th>
      <th>envelope end</th>
      <th>hmm acc</th>
      <th>hmm name</th>
      <th>type</th>
      <th>hmm start</th>
      <th>hmm end</th>
      <th>hmm length</th>
      <th>bit score</th>
      <th>E-value</th>
      <th>clan</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>4798</th>
      <td>A0A096LPK9</td>
      <td>33</td>
      <td>303</td>
      <td>31</td>
      <td>305</td>
      <td>PF13853</td>
      <td>7tm_4</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>
      <td>3</td>
      <td>278</td>
      <td>280</td>
      <td>151.2</td>
      <td>1.200000e-40</td>
      <td>CL0192</td>
    </tr>
    <tr>
      <th>9231</th>
      <td>A0A0G2JMP0</td>
      <td>32</td>
      <td>305</td>
      <td>31</td>
      <td>307</td>
      <td>PF13853</td>
      <td>7tm_4</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>
      <td>2</td>
      <td>278</td>
      <td>280</td>
      <td>130.1</td>
      <td>3.200000e-34</td>
      <td>CL0192</td>
    </tr>
    <tr>
      <th>9443</th>
      <td>A0A0G2JNH3</td>
      <td>33</td>
      <td>303</td>
      <td>31</td>
      <td>305</td>
      <td>PF13853</td>
      <td>7tm_4</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>
      <td>3</td>
      <td>278</td>
      <td>280</td>
      <td>147.6</td>
      <td>1.500000e-39</td>
      <td>CL0192</td>
    </tr>
    <tr>
      <th>11021</th>
      <td>A0A0X1KG70</td>
      <td>32</td>
      <td>305</td>
      <td>31</td>
      <td>307</td>
      <td>PF13853</td>
      <td>7tm_4</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>
      <td>2</td>
      <td>278</td>
      <td>280</td>
      <td>128.8</td>
      <td>8.300000e-34</td>
      <td>CL0192</td>
    </tr>
    <tr>
      <th>14495</th>
      <td>A0A286YEU6</td>
      <td>34</td>
      <td>304</td>
      <td>30</td>
      <td>307</td>
      <td>PF13853</td>
      <td>7tm_4</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x107e...</td>
      <td>4</td>
      <td>276</td>
      <td>280</td>
      <td>124.2</td>
      <td>2.100000e-32</td>
      <td>CL0192</td>
    </tr>
  </tbody>
</table>
</div>




```python
seqs = [dfind.loc[row['seq id']]['seq'][row['envelope start']:row['envelope end']]
        for ind, row in df_pfam_filtered[df_pfam_filtered['hmm acc'] == 'PF13853'].iterrows()]
```


```python

```
#### plot_loglog.py

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('../peptidome.mplstyle')

import sys
sys.path.append('..')
from lib import *

fig, ax = plt.subplots(figsize=(2.5, 1.75))
df = pd.read_csv('data/mutualinformation_uniquedomains.csv')
ax.plot(df['gaps']+1, df['mutualinformation'], 'o', ms=1, label='data')
#err = ax.errorbar(df['gaps']+1, df['mutualinformation'],
#        2*df['mutualinformation_std'], fmt='_', label=species, ms=3)
print(np.sum(df['mutualinformation']))
print(np.sum(2*df['mutualinformation'].iloc[:4]))
d = np.linspace(1, 500)
l, = ax.plot(d, 0.02*d**-.5, zorder=10, c='k', label=r'$d^{-1/2}$')
ax.legend(loc='upper right', ncol=1)
ax.set_ylim(0.0004, 0.05)
ax.set_xlim(0.9, 501.0)
ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlabel('Distance d')
ax.set_ylabel('Mutual information in bits')
fig.tight_layout()
fig.savefig('loglog.pdf')
plt.show()

```
#### run.py

```python
import numpy as np
import pandas as pd

import sys
sys.path.append('..')
from lib import *

#with open('data/human_replaced.txt', 'r') as f:
#    seqs = [s.strip() for s in f.readlines()]
#proteome = seqs
seqs = list(pd.read_csv('data/human_uniquedomains.csv')['seq'])

mutualinformation = []
mutualinformation_std = []
gaps = np.arange(0, 201, 1)
for gap in gaps:
    df2 = Counter(seqs, k=2, gap=gap).to_df(norm=False, clean=True)
    mi = calc_mi(df2)
    mutualinformation.append(mi)
    mi_std = calc_mi_std(seqs, gap)[1]
    mutualinformation_std.append(mi_std)
    print(gap, mi, mi_std)

df = pd.DataFrame.from_dict(dict(gaps=gaps, mutualinformation=mutualinformation,
                                 mutualinformation_std=mutualinformation_std))
df.to_csv('data/mutualinformation_uniquedomains.csv')

```
#### plot.py

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('../peptidome.mplstyle')

import sys
sys.path.append('..')
from lib import *

fig, ax = plt.subplots()
df = pd.read_csv('../mivsdist/data/mutualinformation-Human.csv')
ax.plot(df['gaps']+1, df['mutualinformation'], label='all')
df = pd.read_csv('data/mutualinformation_nozf.csv')
ax.plot(df['gaps']+1, df['mutualinformation'], label='excluding zincfinger')
df = pd.read_csv('data/mutualinformation_uniquedomains.csv')
ax.plot(df['gaps']+1, df['mutualinformation'], label='unique domains')
#df = pd.read_csv('data/mutualinformation_replaced.csv')
#ax.plot(df['gaps']+1, df['mutualinformation'], label='replaced')
ax.legend()
ax.set_ylim(0.0)
ax.set_xlim(1.0, 200.0)
ax.set_xscale('log')
ax.set_xlabel('Distance')
ax.set_ylabel('Mutual information in bits')
fig.tight_layout()
fig.savefig('main.png')
plt.show()

```
