---
layout: post
title: Mutual information as a function of distance
---

An analysis for different proteomes. It seems like the mutual information decays as a power-law with distance (Refer to Lin, Tegmark Entropy 2017 for reference). 

{% include post-image-gallery.html filter="mivsdist/" %}

### Code 
#### backgroundmi.ipynb

```python
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import sys
sys.path.append('..')
from lib import *
```


```python
data = np.load('../aafreqpca/data/data.npz')
aa = data['human']
```

## Mutual information between amino acid and protein

X = amino acid, Y = protein. We have $\operatorname{I}(X;Y) = \mathbb{E}_Y\left[D_\text{KL}\!\left(p_{X|Y} \parallel p_X\right)\right]$, and P(Y) = 1/number of proteins.


```python
meanfreq = aa.mean(axis=0)
np.sum(aa*np.log2(aa/meanfreq))/aa.shape[0]
```




    0.09741991137385587



## Mutual information in shuffled data

$\sum_i \sum_j < f_i^p f_j^p> \log_2 \frac{< f_i^p f_j^p>}{ <f_i^p> <f_j^p>}$


```python
meanfreqs = np.einsum('i,j->ij', meanfreq, meanfreq)
```


```python
freqss = np.einsum('ki,kj->ij', aa, aa)/aa.shape[0]
np.sum(freqss * np.log2(freqss/meanfreqs))
```




    0.0014361921190761484



## Kullback-Leibler Divergences between protein amino acid distributions and the background


```python
freqs = np.einsum('ki,kj->kij', aa, aa)
devs = np.sum(np.sum(freqs * np.log2(freqs/meanfreqs[np.newaxis, :, :]), axis=1), axis=1)
```


```python
plt.hist(devs.flatten(), bins=np.linspace(0.0, np.percentile(devs, 99), 150));
plt.xlabel('DKL(protein aa frequencies | average aa frequencies)')
plt.ylabel('Histogram');
```


![png](notebook_files/backgroundmi_9_0.png)



```python

```
#### plot_total.py

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('../peptidome.mplstyle')

import sys
sys.path.append('..')
from lib import *


fig, ax = plt.subplots()
for species in ['Human', 'Yeast']:
    df = pd.read_csv('data/mutualinformation-%s.csv'%species)
    print(df['gaps'].max())
    ks = np.arange(1, df['gaps'].max())
    totalmi = [np.sum(2*df['mutualinformation'].iloc[:k]) for k in ks]
    totalmistd = np.array([2*np.sum(df['mutualinformation_std'].iloc[:k]**2)**.5 for k in ks])
    err = ax.errorbar(ks, totalmi, 2*totalmistd,  fmt='o', label=species, ms=1)
d = np.linspace(1, 200)
#ax.plot(d, 0.005*np.exp(-d/50), zorder=10, label='exponential ($d_c=50$)')
#l, = ax.plot(d, 0.02*d**-.5, zorder=10, c='k', label=r'power law ($-1/2$)')
#ax.plot(d, 0.009*d**-.5, zorder=10, c=l.get_color())
ax.legend(loc='upper right', ncol=2)
#ax.set_ylim(0.0004, 0.05)
#ax.set_xlim(0.95, 201.0)
ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlabel('Distance')
ax.set_ylabel('Mutual information in bits')
fig.tight_layout()
fig.savefig('total.png')
plt.show()

```
#### plot_loglog.py

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('../peptidome.mplstyle')

import sys
sys.path.append('..')
from lib import *

fig, ax = plt.subplots()
for species in ['Human', 'Yeast']:
    df = pd.read_csv('data/mutualinformation-%s.csv'%species)
    err = ax.errorbar(df['gaps']+1, df['mutualinformation'],
            2*df['mutualinformation_std'], fmt='_', label=species, ms=3)
    print(np.sum(df['mutualinformation'].iloc[:200]))
    print(np.sum(2*df['mutualinformation'].iloc[:4]))
#    ax.plot(df['gaps']+1, df['shuffledmutualinformation'], '_', ms=3,
#            label=('shuffled') if species == 'Human' else '', c=err.lines[0].get_color())
d = np.linspace(1, 200)
#ax.plot(d, 0.005*np.exp(-d/50), zorder=10, label='exponential ($d_c=50$)')
l, = ax.plot(d, 0.02*d**-.5, zorder=10, c='k', label=r'power law ($-1/2$)')
ax.plot(d, 0.009*d**-.5, zorder=10, c=l.get_color())
ax.legend(loc='upper right', ncol=2)
ax.set_ylim(0.0004, 0.05)
ax.set_xlim(0.95, 201.0)
ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlabel('Distance')
ax.set_ylabel('Mutual information in bits')
fig.tight_layout()
fig.savefig('loglog.png')
plt.show()

```
#### run.py

```python
import numpy as np
import pandas as pd

import sys
sys.path.append('..')
from lib import *

run = int(sys.argv[1])-1
names = ['Human', 'Mouse', 'Yeast', 'Viruses']
name = names[run]
if name == 'Viruses':
    proteome = datadir + 'human-viruses-uniref90_nohiv.fasta'
else:
    proteome = proteome_path(name)
print(name, proteome)

mutualinformation = []
mutualinformation_std = []
shuffled_mutualinformation = []
gaps = np.arange(0, 201, 1)
for gap in gaps:
    seqs = [s for s in fasta_iter(proteome, returnheader=False)]
    df2 = Counter(seqs, k=2, gap=gap).to_df(norm=False, clean=True)
    mi = calc_mi(df2)
    mutualinformation.append(mi)
    mi_std = calc_mi_std(seqs, gap)
    mutualinformation_std.append(mi_std)

    # calculate shuffled mi
    iterable = scrambled(fasta_iter(proteome, returnheader=False))
    df2 = Counter(iterable, k=2, gap=gap).to_df(norm=False, clean=True)
    shuffledmi = calc_mi(df2)
    shuffled_mutualinformation.append(shuffledmi)
    print(gap, mi, mi_std, shuffledmi)

df = pd.DataFrame.from_dict(dict(gaps=gaps, mutualinformation=mutualinformation,
                                 mutualinformation_std=mutualinformation_std,
                                 shuffledmutualinformation=shuffled_mutualinformation))
df.to_csv('data/mutualinformation-%s.csv'%name)

```
#### plot.py

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('../peptidome.mplstyle')

import sys
sys.path.append('..')
from lib import *

for species in ['Human', 'Mouse', 'Yeast', 'Viruses']:
    fig, ax = plt.subplots()
    df = pd.read_csv('data/mutualinformation-%s.csv'%species)
    l, = ax.plot(df['gaps']+1, df['mutualinformation'], lw=.8)
    ax.fill_between(df['gaps']+1,
                    df['mutualinformation']-df['mutualinformation_std'],
                    df['mutualinformation']+df['mutualinformation_std'],
                    color=l.get_color(), alpha=.3)
    #ax.errorbar(df['gaps']+1, df['mutualinformation'],
    #        2*df['mutualinformation_std'], label='data')
    ax.plot(df['gaps']+1, df['shuffledmutualinformation'], label='shuffled', lw=.8)
    ax.legend()
    ax.set_ylim(0.0, 0.0185)
    ax.set_xlim(0.95, 201.0)
    ax.set_xscale('log')
    ax.set_xlabel('Distance')
    ax.set_ylabel('Mutual information in bits')
    fig.tight_layout()
    if species == 'Human':
        fig.savefig('main.png')
        plt.show()
    else:
        fig.savefig('%s.png'%species)

```
