---
layout: post
title: How much confounding is there due to shared protein domains?
---

An analysis based on Pfam domain assignments. As expected the dominant peaks in the mutual information vanish once we exclude all proteins containing Zinc finger motifs. We can also use the domain information in various ways to filter the protein data to more generally assess how much shared protein domains affect the mutual information. One sensible way to do so is to keep just one randmly chosen protein per protein domain family. This also removes the Zinc finger peak, but otherwise the decay of mutual information is broadly similar. 

{% include post-image-gallery.html filter="pfam/" %}

### Code 
#### pfam.ipynb

```python
import sys
sys.path.append('..')
import random
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import projgrad
plt.style.use('../peptidome.mplstyle')

from lib import *
```

Pfam data for human:

https://pfam.xfam.org/proteome/9606

Sequence coverage 71% and residue coverage 45%


```python
columnsstr = 'seq id> <alignment start> <alignment end> <envelope start> <envelope end> <hmm acc> <hmm name> <type> <hmm start> <hmm end> <hmm length> <bit score> <E-value> <clan'
columns = columnsstr.split('> <')
df_pfam = pd.read_csv(datadir+'9606_pfam.tsv.gz', sep='\t', skiprows=3, names=columns)
```


```python
df = load_proteome_as_df('Human')
df = df[~(df['Gene']=='')]
seqids_set = set(df['Accession'])
```


```python
# pfam contains info for complete proteome for 74k sequences for human
# filter on sequence ids found in reference proteome
df_pfam = df_pfam[df_pfam['seq id'].isin(seqids_set)]
```


```python
len(set(df['Accession'])-set(df_pfam['seq id'])), len(set(df['Accession'])), len(set(df_pfam['seq id']))
```




    (1737, 20480, 18743)




```python
pfam_multiplicities = df_pfam['hmm acc'].value_counts()
#pfams_ordered = pfams_ordered.reset_index()
pfam_multiplicities.head()
```




    PF00096    5390
    PF00400     751
    PF00028     693
    PF00041     612
    PF07679     566
    Name: hmm acc, dtype: int64




```python
pfam_multiplicities = df_pfam['hmm name'].value_counts()
#pfams_ordered = pfams_ordered.reset_index()
pfam_multiplicities.head()
```




    zf-C2H2     5390
    WD40         751
    Cadherin     693
    fn3          612
    I-set        566
    Name: hmm name, dtype: int64




```python
plt.hist(pfam_multiplicities, bins=np.logspace(0, 4))
plt.xscale('log')
plt.yscale('log')
```

    The history saving thread hit an unexpected error (OperationalError('database is locked',)).History will not be written to the database.



![png](notebook_files/pfam_8_1.png)


## Zincfingers


```python
zfs = df_pfam[df_pfam['hmm acc'] == 'PF00096']
zfs.head()
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seq id</th>
      <th>alignment start</th>
      <th>alignment end</th>
      <th>envelope start</th>
      <th>envelope end</th>
      <th>hmm acc</th>
      <th>hmm name</th>
      <th>type</th>
      <th>hmm start</th>
      <th>hmm end</th>
      <th>hmm length</th>
      <th>bit score</th>
      <th>E-value</th>
      <th>clan</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>4169</th>
      <td>A0A087X1K6</td>
      <td>195</td>
      <td>219</td>
      <td>195</td>
      <td>219</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>24.0</td>
      <td>0.093</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>4170</th>
      <td>A0A087X1K6</td>
      <td>225</td>
      <td>247</td>
      <td>225</td>
      <td>247</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>23.6</td>
      <td>0.120</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>4171</th>
      <td>A0A087X1K6</td>
      <td>165</td>
      <td>189</td>
      <td>165</td>
      <td>189</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>18.3</td>
      <td>5.800</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>6983</th>
      <td>A0A0A6YYC7</td>
      <td>372</td>
      <td>394</td>
      <td>372</td>
      <td>394</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>19.9</td>
      <td>1.800</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>6984</th>
      <td>A0A0A6YYC7</td>
      <td>401</td>
      <td>422</td>
      <td>400</td>
      <td>422</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>2</td>
      <td>23</td>
      <td>23</td>
      <td>19.7</td>
      <td>2.100</td>
      <td>CL0361</td>
    </tr>
  </tbody>
</table>
</div>




```python
dfind = df.set_index('Accession')
seq_zf = [dfind.loc[row['seq id']]['Sequence'][row['envelope start']:row['envelope end']] for ind, row in zfs.iterrows()]
seq_zf[:5]
```




    ['ACSWQDCNKKFARSDELARHYRTH',
     'SCPICEKRFMRSDHLTKHARRH',
     'KCHYAGCEKVYGKSSHLKAHLRTH',
     'ICEYCARAFKSSHNLAVHRMIH',
     'QCEICGFTCRQKASLNWHMKKH']




```python
df_nozf = df[~df['Accession'].isin(set(zfs['seq id']))]
df_nozf.to_csv('data/human_nozf.csv', index=False)
```


```python
seqid = "A8MTY0"
zfs_seq = zfs[zfs['seq id']==seqid]
zfs_seq.sort_values('envelope start').head()
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seq id</th>
      <th>alignment start</th>
      <th>alignment end</th>
      <th>envelope start</th>
      <th>envelope end</th>
      <th>hmm acc</th>
      <th>hmm name</th>
      <th>type</th>
      <th>hmm start</th>
      <th>hmm end</th>
      <th>hmm length</th>
      <th>bit score</th>
      <th>E-value</th>
      <th>clan</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>16596</th>
      <td>A8MTY0</td>
      <td>170</td>
      <td>192</td>
      <td>170</td>
      <td>192</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>20.8</td>
      <td>0.940</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16597</th>
      <td>A8MTY0</td>
      <td>226</td>
      <td>248</td>
      <td>226</td>
      <td>248</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>20.6</td>
      <td>1.100</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16588</th>
      <td>A8MTY0</td>
      <td>282</td>
      <td>304</td>
      <td>282</td>
      <td>304</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>25.2</td>
      <td>0.037</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16587</th>
      <td>A8MTY0</td>
      <td>310</td>
      <td>332</td>
      <td>310</td>
      <td>332</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>26.4</td>
      <td>0.016</td>
      <td>CL0361</td>
    </tr>
    <tr>
      <th>16591</th>
      <td>A8MTY0</td>
      <td>338</td>
      <td>360</td>
      <td>338</td>
      <td>360</td>
      <td>PF00096</td>
      <td>zf-C2H2</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89b8...</td>
      <td>1</td>
      <td>23</td>
      <td>23</td>
      <td>23.5</td>
      <td>0.130</td>
      <td>CL0361</td>
    </tr>
  </tbody>
</table>
</div>




```python
proteinatlas = pd.read_csv('../../data/proteinatlas.tsv.zip', sep='\t', index_col=0)
tissues = [c for c in proteinatlas.columns if 'Tissue' in c]
proteinatlas[tissues] = proteinatlas[tissues].div(proteinatlas[tissues].sum(axis=0), axis=1)
proteinatlas.fillna(0, inplace=True)
proteinatlas.head()
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene synonym</th>
      <th>Ensembl</th>
      <th>Gene description</th>
      <th>Uniprot</th>
      <th>Chromosome</th>
      <th>Position</th>
      <th>Protein class</th>
      <th>Evidence</th>
      <th>HPA evidence</th>
      <th>UniProt evidence</th>
      <th>...</th>
      <th>Brain RNA - amygdala [NX]</th>
      <th>Brain RNA - basal ganglia [NX]</th>
      <th>Brain RNA - cerebellum [NX]</th>
      <th>Brain RNA - cerebral cortex [NX]</th>
      <th>Brain RNA - hippocampal formation [NX]</th>
      <th>Brain RNA - hypothalamus [NX]</th>
      <th>Brain RNA - midbrain [NX]</th>
      <th>Brain RNA - olfactory region [NX]</th>
      <th>Brain RNA - pons and medulla [NX]</th>
      <th>Brain RNA - thalamus [NX]</th>
    </tr>
    <tr>
      <th>Gene</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TSPAN6</th>
      <td>T245, TM4SF6, TSPAN-6</td>
      <td>ENSG00000000003</td>
      <td>Tetraspanin 6</td>
      <td>O43657</td>
      <td>X</td>
      <td>100627109-100639991</td>
      <td>Predicted intracellular proteins, Predicted me...</td>
      <td>Evidence at protein level</td>
      <td>Evidence at transcript level</td>
      <td>Evidence at protein level</td>
      <td>...</td>
      <td>7.0</td>
      <td>6.6</td>
      <td>2.9</td>
      <td>4.9</td>
      <td>7.4</td>
      <td>7.3</td>
      <td>6.9</td>
      <td>4.1</td>
      <td>34.2</td>
      <td>6.6</td>
    </tr>
    <tr>
      <th>TNMD</th>
      <td>BRICD4, ChM1L, myodulin, TEM, tendin</td>
      <td>ENSG00000000005</td>
      <td>Tenomodulin</td>
      <td>Q9H2S6</td>
      <td>X</td>
      <td>100584802-100599885</td>
      <td>Predicted membrane proteins</td>
      <td>Evidence at protein level</td>
      <td>Evidence at transcript level</td>
      <td>Evidence at protein level</td>
      <td>...</td>
      <td>0.1</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.1</td>
      <td>0.9</td>
      <td>0.1</td>
      <td>0.6</td>
      <td>0.0</td>
      <td>0.6</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>DPM1</th>
      <td>CDGIE, MPDS</td>
      <td>ENSG00000000419</td>
      <td>Dolichyl-phosphate mannosyltransferase subunit...</td>
      <td>O60762</td>
      <td>20</td>
      <td>50934867-50958555</td>
      <td>Disease related genes, Enzymes, Plasma protein...</td>
      <td>Evidence at protein level</td>
      <td>Evidence at transcript level</td>
      <td>Evidence at protein level</td>
      <td>...</td>
      <td>20.2</td>
      <td>30.0</td>
      <td>22.5</td>
      <td>35.1</td>
      <td>22.2</td>
      <td>21.5</td>
      <td>21.6</td>
      <td>17.2</td>
      <td>31.4</td>
      <td>21.7</td>
    </tr>
    <tr>
      <th>SCYL3</th>
      <td>PACE-1, PACE1</td>
      <td>ENSG00000000457</td>
      <td>SCY1 like pseudokinase 3</td>
      <td>Q8IZE3</td>
      <td>1</td>
      <td>169849631-169894267</td>
      <td>Enzymes, Predicted intracellular proteins</td>
      <td>Evidence at protein level</td>
      <td>Evidence at transcript level</td>
      <td>Evidence at protein level</td>
      <td>...</td>
      <td>8.5</td>
      <td>10.6</td>
      <td>10.5</td>
      <td>11.9</td>
      <td>7.3</td>
      <td>6.5</td>
      <td>9.3</td>
      <td>7.8</td>
      <td>9.5</td>
      <td>8.9</td>
    </tr>
    <tr>
      <th>C1orf112</th>
      <td>FLJ10706</td>
      <td>ENSG00000000460</td>
      <td>Chromosome 1 open reading frame 112</td>
      <td>Q9NSG2</td>
      <td>1</td>
      <td>169662007-169854080</td>
      <td>Predicted intracellular proteins</td>
      <td>Evidence at protein level</td>
      <td>Evidence at transcript level</td>
      <td>Evidence at protein level</td>
      <td>...</td>
      <td>3.3</td>
      <td>4.9</td>
      <td>8.4</td>
      <td>3.6</td>
      <td>2.7</td>
      <td>2.4</td>
      <td>5.0</td>
      <td>1.5</td>
      <td>4.5</td>
      <td>5.9</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 226 columns</p>
</div>




```python
excluded_genes = set(df[df['Accession'].isin(set(zfs['seq id']))]['Gene'])
plt.hist(proteinatlas.loc[excluded_genes].filter(regex='Tissue').sum());
```

    /home/amayer/.conda/envs/py3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: 
    Passing list-likes to .loc or [] with any missing label will raise
    KeyError in the future, you can use .reindex() as an alternative.
    
    See the documentation here:
    https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike
      



![png](notebook_files/pfam_15_1.png)



```python
multiple_domains_hmm = set(pfam_multiplicities[pfam_multiplicities > 10].index)
```


```python
df_multiple = df_pfam[df_pfam['hmm acc'].isin(multiple_domains_hmm)]
df_multiple.head()
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>seq id</th>
      <th>alignment start</th>
      <th>alignment end</th>
      <th>envelope start</th>
      <th>envelope end</th>
      <th>hmm acc</th>
      <th>hmm name</th>
      <th>type</th>
      <th>hmm start</th>
      <th>hmm end</th>
      <th>hmm length</th>
      <th>bit score</th>
      <th>E-value</th>
      <th>clan</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>16</th>
      <td>A0A024R161</td>
      <td>49</td>
      <td>115</td>
      <td>49</td>
      <td>120</td>
      <td>PF00226</td>
      <td>DnaJ</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x89d6...</td>
      <td>1</td>
      <td>57</td>
      <td>63</td>
      <td>56.3</td>
      <td>6.600000e-12</td>
      <td>CL0392</td>
    </tr>
    <tr>
      <th>17</th>
      <td>A0A024R161</td>
      <td>108</td>
      <td>152</td>
      <td>96</td>
      <td>153</td>
      <td>PF00631</td>
      <td>G-gamma</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0x8ba7...</td>
      <td>20</td>
      <td>67</td>
      <td>68</td>
      <td>39.3</td>
      <td>1.200000e-06</td>
      <td>No_clan</td>
    </tr>
    <tr>
      <th>175</th>
      <td>A0A075B6H5</td>
      <td>39</td>
      <td>128</td>
      <td>39</td>
      <td>130</td>
      <td>PF07686</td>
      <td>V-set</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0xf3db...</td>
      <td>1</td>
      <td>91</td>
      <td>109</td>
      <td>40.8</td>
      <td>5.100000e-07</td>
      <td>CL0011</td>
    </tr>
    <tr>
      <th>176</th>
      <td>A0A075B6H7</td>
      <td>26</td>
      <td>110</td>
      <td>26</td>
      <td>115</td>
      <td>PF07686</td>
      <td>V-set</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0xf3db...</td>
      <td>1</td>
      <td>91</td>
      <td>109</td>
      <td>63.3</td>
      <td>5.300000e-14</td>
      <td>CL0011</td>
    </tr>
    <tr>
      <th>177</th>
      <td>A0A075B6H8</td>
      <td>28</td>
      <td>111</td>
      <td>28</td>
      <td>117</td>
      <td>PF07686</td>
      <td>V-set</td>
      <td>PfamLive::Result::SequenceOntology=HASH(0xf3db...</td>
      <td>1</td>
      <td>91</td>
      <td>109</td>
      <td>42.3</td>
      <td>1.700000e-07</td>
      <td>CL0011</td>
    </tr>
  </tbody>
</table>
</div>




```python
multiple_domains_seqid = set(df_multiple['seq id'])
```


```python
len(seqids_set-multiple_domains_seqid), len(seqids_set), len(multiple_domains_seqid)
```




    (9115, 21080, 11965)




```python
seqs = []
with open('data/human_replaced.csv', 'w') as f:
    f.write('Accession' + ',' + 'Sequence'+ '\n')
    for ind, row in df.iterrows():
        seqid, seq = row['Accession'], row['Sequence']
        pfam = df_multiple[df_multiple['seq id']==seqid]
        if pfam.shape[0] == 0:
            newseq = seq
        else:
            newseq = ''
            start = 0
            pfam = pfam.sort_values('envelope start')
            for ind, row in pfam.iterrows():
                end = row['envelope start']
                newseq += seq[start:end]
                newseq += 'X'*(row['envelope end']-row['envelope start'])
                start = row['envelope end']
            end = row['envelope end']
            newseq += seq[end:]
        seqs.append(newseq)
        f.write(seqid + ',' + newseq + '\n')
```


```python
pfam_nproteins = df_pfam.groupby(['seq id', 'hmm name']).mean().reset_index()['hmm name'].value_counts()
pfam_nproteins.head(n=20)
```




    zf-C2H2        681
    7tm_4          424
    V-set          421
    KRAB           368
    Pkinase        347
    7tm_1          287
    WD40           219
    Homeodomain    219
    RRM_1          207
    Ank_2          199
    PH             180
    LRR_8          171
    fn3            136
    Ras            136
    I-set          133
    Ig_3           132
    Pkinase_Tyr    129
    BTB            129
    C2             123
    Trypsin        121
    Name: hmm name, dtype: int64




```python
pfam_nproteins = df_pfam.groupby(['seq id', 'hmm acc']).mean().reset_index()['hmm acc'].value_counts()
pfam_nproteins.head(n=20)
```




    PF00096    681
    PF13853    424
    PF07686    421
    PF01352    368
    PF00069    347
    PF00001    287
    PF00400    219
    PF00046    219
    PF00076    207
    PF12796    199
    PF00169    180
    PF13855    171
    PF00041    136
    PF00071    136
    PF07679    133
    PF13927    132
    PF00651    129
    PF07714    129
    PF00168    123
    PF00089    121
    Name: hmm acc, dtype: int64




```python
plt.hist(pfam_nproteins, bins=np.logspace(0, 3))
plt.xscale('log')
plt.yscale('log')
```


![png](notebook_files/pfam_23_0.png)



```python
ntop = 10
top = set(df_pfam[df_pfam['hmm acc'].isin(set(pfam_nproteins.iloc[:ntop].index))]['seq id'])
df_notop = df[~df['Accession'].isin(top)]
df_notop.to_csv('data/human_notop%i.csv'%ntop, index=False)
len(top), sum(pfam_nproteins.iloc[:ntop]), '%e, %e' %(np.sum(df[~df['Accession'].isin(top)]['Sequence'].apply(len)), np.sum(df['Sequence'].apply(len)))
```




    (3018, 3372, '9.766670e+06, 1.137888e+07')




```python
excluded_genes = set(df[df['Accession'].isin(top)]['Gene'])
plt.hist(proteinatlas.loc[excluded_genes].filter(regex='Tissue').sum());
```

    /home/amayer/.conda/envs/py3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: 
    Passing list-likes to .loc or [] with any missing label will raise
    KeyError in the future, you can use .reindex() as an alternative.
    
    See the documentation here:
    https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike
      



![png](notebook_files/pfam_25_1.png)



```python
# keep one random protein per family
remove = set()
#for pfam in pfam_nproteins.iloc[:ntop].index:
for pfam in pfam_nproteins[pfam_nproteins > 10].index:
    family = set(df_pfam[df_pfam['hmm acc'] == pfam]['seq id'].unique())
    family = family - set(random.sample(family, 10))
    remove = remove | family
len(remove)
```




    8786




```python
'%e, %e' %(np.sum(df[~df['Accession'].isin(remove)]['Sequence'].apply(len)), np.sum(df['Sequence'].apply(len)))
```




    '5.574679e+06, 1.137888e+07'




```python
df_domains = df[~df['Accession'].isin(remove)]
df_domains.to_csv('data/human_downsampled.csv', index=False)
```

## Olfactory receptors


```python
olfactory_receptors = df_pfam[df_pfam['hmm acc'] == 'PF13853']
olfactory_receptors.shape[0]
```




    424




```python
seqs = [dfind.loc[row['seq id']]['Sequence'][row['envelope start']:row['envelope end']]
        for ind, row in olfactory_receptors.iterrows()]
```


```python

```
#### plot_loglog.py

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('../peptidome.mplstyle')

import sys
sys.path.append('..')
from lib import *

fig, ax = plt.subplots(figsize=(2.5, 1.75))
df = pd.read_csv('data/mutualinformation_nozf.csv')
ax.plot(df['gaps']+1, df['mutualinformation'], 'o', ms=1, label='data')
#err = ax.errorbar(df['gaps']+1, df['mutualinformation'],
#        2*df['mutualinformation_std'], fmt='_', label=species, ms=3)
print(np.sum(df['mutualinformation']))
print(np.sum(2*df['mutualinformation'].iloc[:4]))
d = np.linspace(1, 500)
l, = ax.plot(d, 0.018*d**-.5, zorder=-1, c='k', label=r'$d^{-1/2}$')
ax.legend(loc='upper right', ncol=1)
ax.set_ylim(0.0004, 0.05)
ax.set_xlim(0.9, 501.0)
ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlabel('Distance d')
ax.set_ylabel('Mutual information in bits')
fig.tight_layout()
fig.savefig('loglog.png')
plt.show()

```
#### run.py

```python
import numpy as np
import pandas as pd

import sys
sys.path.append('..')
from lib import *

filtering = 'notop20'
seqs = list(pd.read_csv('data/human_%s.csv'%filtering)['Sequence'])

mutualinformation = []
mutualinformation_std = []
gaps = np.arange(0, 201, 1)
for gap in gaps:
    df2 = Counter(seqs, k=2, gap=gap).to_df(norm=False, clean=True)
    mi = calc_mi(df2)
    mutualinformation.append(mi)
    mi_std = calc_mi_std(seqs, gap)[1]
    mutualinformation_std.append(mi_std)
    print(gap, mi, mi_std)

df = pd.DataFrame.from_dict(dict(gaps=gaps, mutualinformation=mutualinformation,
                                 mutualinformation_std=mutualinformation_std))
df.to_csv('data/mutualinformation_%s.csv'%filtering)

```
#### plot.py

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('../peptidome.mplstyle')

import sys
sys.path.append('..')
from lib import *

fig, ax = plt.subplots()
df = pd.read_csv('../mivsdist/data/mutualinformation-Human.csv')
ax.plot(df['gaps']+1, df['mutualinformation'], label='all')
df = pd.read_csv('data/mutualinformation_nozf.csv')
ax.plot(df['gaps']+1, df['mutualinformation'], label='excluding zincfinger')
df = pd.read_csv('data/mutualinformation_uniquedomains.csv')
ax.plot(df['gaps']+1, df['mutualinformation'], label='unique domains')
#df = pd.read_csv('data/mutualinformation_replaced.csv')
#ax.plot(df['gaps']+1, df['mutualinformation'], label='replaced')
ax.legend()
ax.set_ylim(0.0)
ax.set_xlim(1.0, 200.0)
ax.set_xscale('log')
ax.set_xlabel('Distance')
ax.set_ylabel('Mutual information in bits')
fig.tight_layout()
fig.savefig('main.png')
plt.show()

```
